pkg/registries/registries.go

```diff
@@ -3,17 +3,48 @@ package registries
 import (
 	"fmt"
 	"sort"
+	"strings"
 
 	config "github.com/openshift/api/config/v1"
 )
 
 // mirror represents a mirrored image repo entry in a registry configuration file
 type mirror struct {
+	// host is the mirror image target location. Can include the registry hostname/IP address, port, and namespace path
 	host string
 	// resolveTags indicates to the container runtime if this mirror is allowed to resolve an image tag into a digest
 	resolveTags bool
 }
 
+func newMirror(sourceImageLocation, mirrorImageLocation string, resolveTags bool) mirror {
+	mirrorHost := ""
+	if sourceImageLocation == mirrorImageLocation {
+		// special case if source and mirror are the same
+		mirrorHost = extractHostname(mirrorImageLocation)
+	} else {
+		// truncate the mirror to drop any shared namespaces since containerd automatically appends them on image pull
+		mirrorHost = dropCommonSuffix(sourceImageLocation, mirrorImageLocation)
+	}
+	return mirror{host: mirrorHost, resolveTags: resolveTags}
+}
+
+// dropCommonSuffix drops the common suffix from the second repo, returning only the unique leading URL and namespaces
+func dropCommonSuffix(source, mirror string) string {
+	sourceParts := strings.Split(source, "/")
+	mirrorParts := strings.Split(mirror, "/")
+	uniqueMirrorParts := mirrorParts
+
+	// Process until we hit the end of either repo string
+	for i := 0; i < len(sourceParts) && i < len(mirrorParts); i++ {
+		// Check if suffix piece is equal, starting from the backs of the lists
+		if sourceParts[len(sourceParts)-1-i] == mirrorParts[len(mirrorParts)-1-i] {
+			// Remove common suffix piece
+			uniqueMirrorParts = uniqueMirrorParts[:len(uniqueMirrorParts)-1]
+		}
+	}
```

Comment:
Does this work as expected in the case where there are two strings
`foo1/fah/different/foo2`
`foo1/fah/something/foo2`
Tried this and its giving me `foo1`, when it really should give `foo1/fah` ?

I believe you should break when you hit something that is different.
I suggest doing it this way instead:
```
	// Process until we hit the end of either repo string
	var index int
	for index = 0; index < len(sourceParts) && index < len(mirrorParts); index++ {
		// Check if suffix piece is equal, starting from the backs of the lists
		if sourceParts[len(sourceParts)-1-index] != mirrorParts[len(mirrorParts)-1-index] {
			break
		}
	}

	fmt.Println(strings.Join(mirrorParts[:len(mirrorParts)-1-index], "/"))
```

But does this required that sourceParts and mirrorParts are of equal length, or that one specific one is longer than the other?

Reply:
> when it really should give foo1/fah

it should give `foo1/fah/something` as that is the mirror location minus any common suffixes. Good catch though, you're right it doesn't work as intended in that case!

Reply:
I did a combination of the two approaches to avoid the equal length concern and added some unit tests to prove it works now

Reply:
I recommend moving this logic into a separate helper function so that it can be unit-tested.